src/transport/icmp4.rs

```diff
@@ -0,0 +1,243 @@
+        self.icmp_type as u8
+    }
+    fn get_raw_code(&self) -> u8 {
+        0u8
+    }
+    fn get_checksum(&self) -> u16 {
+        0u16
+    }
+
+    fn get_type_dependent_data(&self) -> u32 {
+        use IcmpV4Type::*;
+        match self.icmp_type {
+            EchoRequest| EchoReply => {
+                if let Some(echo_header) = &self.echo_header {
+                    let seq_be = echo_header.seq.to_be_bytes();
+                    let id_be = echo_header.id.to_be_bytes();
+                    u32::from_be_bytes([seq_be[0], seq_be[1], id_be[0], id_be[1]])
+                } else {
+                    // caller never setup the IcmpEchoHeader
+                    // just assume seq = id = 0
+                    0
+                }
+            },
+            _ => 0 // TODO: fill out other exceptions for this data...
+        }
+    }
+}
+
+///A slice containing an icmp4 header of a network package. Struct allows the selective read of fields in the header.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Icmp4HeaderSlice<'a> {
+    slice: &'a [u8]
+}
+
+impl<'a> Icmp4HeaderSlice<'a> {
+    /// Creates a slice containing an icmp4 header.
+    #[inline]
+    pub fn from_slice(slice: &'a[u8]) -> Result<Icmp4HeaderSlice<'a>, ReadError> {
+        //check length
+        use crate::ReadError::*;
+        if slice.len() < IcmpV4Header::SERIALIZED_SIZE {
+            return Err(UnexpectedEndOfSlice(IcmpV4Header::SERIALIZED_SIZE));
+        }
+
+        //done
+        Ok(Icmp4HeaderSlice{
+            // SAFETY:
+            // Safe as slice length is checked to be at least
+            // IcmpV4Header::SERIALIZED_SIZE (8) before this.
+            slice: unsafe {
+                from_raw_parts(
+                    slice.as_ptr(),
+                    IcmpV4Header::SERIALIZED_SIZE
+                )
+            }
+        })
+    }
+
+
+    fn parse_echo_header(&self) -> Option<IcmpEchoHeader> {
+        use IcmpV4Type::*;
+        if let Ok(icmp_type) = self.icmp_type() {
+            if (icmp_type == EchoReply) || (icmp_type == EchoRequest) {
+                let seq: u16 = u16::from_be_bytes([self.slice[4], self.slice[5]]);
+                let id: u16 = u16::from_be_bytes([self.slice[6], self.slice[7]]);
+                return Some(IcmpEchoHeader{
+                    seq,
+                    id,
+                });
+            }
+        }
+        None
+    }
+
+    /// Decode all the fields and copy the results to a UdpHeader struct
+    #[inline]
+    pub fn to_header(&self) -> Result<IcmpV4Header, ValueError> {
+        Ok(IcmpV4Header {
+            icmp_type: self.icmp_type()?,
+            icmp_code: self.icmp_code()?,
+            icmp_chksum: self.icmp_chksum()?,
+            echo_header: self.parse_echo_header(),
+        })
+    }
+
+    pub fn icmp_type(&self) -> Result<IcmpV4Type, ValueError> {
+        // already checked slice len in ::from_slice()
+        IcmpV4Type::from(self.slice[0])
+    }
```

Comment:
I would directly return the u8.

In the past I also always tried to setup some enum values, but it was ultimately more trouble then it was worth. Keeping an enum up-to-date is almost impossible and then there is the chance that a valid packet gets rejected because of an unknown enum value. It is less troublesome to allow all values for which we are not 100% sure if they are invalid.

Then there is no possible error case and the u8 can be returned without being wrapped in a Result (+ inline).

```suggestion
    #[inline]
    pub fn icmp_type(&self) -> u8 {
        // SAFETY:
        // Safe as the contructor checks that the slice has
        // at least the length of IcmpV4Header::SERIALIZED_SIZE (8).
        unsafe {
            *self.slice.get_unchecked(0)
        }
    }
```